gtsfm/densify/patchmatchnet_data.py

```diff
@@ -0,0 +1,259 @@
+            #   Initially the intrinsics is scaled to fit the smallest image size
+            intrinsics[:2, :] /= 2 ** NUM_PATCHMATCHNET_STAGES
+
+            for i in range(self._num_stages):
+                imgs[i].append(
+                    cv2.resize(
+                        np_img,
+                        (self._cropped_w // (2 ** i), self._cropped_h // (2 ** i)),
+                        interpolation=cv2.INTER_LINEAR,
+                    )
+                )
+                proj_mat = extrinsics.copy()
+                intrinsics[:2, :] *= 2.0
+                proj_mat[:3, :4] = intrinsics @ proj_mat[:3, :4]
+                # For the next stage, the image size is doubled, so the intrinsics should also doubled.
+                proj_mats[-1 - i].append(proj_mat)
+
+        imgs_dict = {}
+        proj_dict = {}
+        for i in range(self._num_stages):
+            # reshaping the images from (B, H, W, C) to (B, C, H, W)
+            imgs_dict[f"stage_{i}"] = np.stack(imgs[i]).transpose([0, 3, 1, 2])
+            proj_dict[f"stage_{i}"] = np.stack(proj_mats[i])
+
+        return {
+            "idx": index,
+            "imgs": imgs_dict,  # N*3*H0*W0
+            "proj_matrices": proj_dict,  # N*4*4
+            "depth_min": self._depth_ranges[index, 0],  # scalar
+            "depth_max": self._depth_ranges[index, 1],  # scalar
+            "filename": "{}/" + f"{ref_key:0>8}" + "{}",
+        }
+
+    def get_packed_pairs(self) -> List[Tuple[int, List[int]]]:
+        """Pack view pair data in the form of (reference view index, source view indices) to fit with inference methods
+
+        Returns:
+            List[Tuple[int, List[int]]]: packed pair data in (reference view index, source view indices), the length of
+                source view indices is num_views
+        """
+        packed_pairs = []
+        for idx in range(self._num_images):
+            ref_view = idx
+            src_views = self._pairs[idx].tolist()
+            packed_pairs.append((ref_view, src_views))
+        return packed_pairs
+
+    def get_camera_params(self, index: int) -> Tuple[np.ndarray, np.ndarray]:
+        """Get camera intrinsics and extrinsics parameters by image(or view) index
+
+        Args:
+            index (int): image(or view) index
+
+        Returns:
+            Tuple[np.ndarray, np.ndarray]: (intrinsics (3, 3), extrinsics (4, 4)) of the input image(or view) index
+        """
+        cam_key = self._patchmatchnet_idx_to_camera_idx[index]
+        intrinsics = self._sfm_result.get_camera(cam_key).calibration().K().copy()
+        extrinsics = self._sfm_result.get_camera(cam_key).pose().inverse().matrix().copy()
+        return (intrinsics, extrinsics)
+
+    def get_image(self, index: int) -> np.ndarray:
+        """Get preprocessed image by image(or view) index
+
+        Args:
+            index (int): image(or view) index
+
+        Returns:
+            np.ndarray: preprocessed image, (cropped_h, cropped_w)
```

Comment:
Prefer to say
‘Preprocessed image, of shape (cropped_h, cropped_w, 3)

is there a channel dim?